File: src/main/java/andrew/solution/MergeSortArray.java
package andrew.solution;

/**
 * @author Andrew Tse
 * @date 2019-06-04
 */
public class MergeSortArray {
    public static void main(String[] args) {
        int[] nums1 = {1, 2, 3, 0, 0, 0};
        int[] nums2 = {2, 5, 6};
        int m = 3;
        int n = 3;
        merge(nums1, m, nums2, n);
        for (int value : nums1) {
            System.out.println(value);
        }
    }

    /**
     * 合并两个有序数组
     * 思路：从尾部开始归并排序
     *
     * @param nums1
     * @param m
     * @param nums2
     * @param n
     */
    private static void merge(int[] nums1, int m, int[] nums2, int n) {
        int i = m + n - 1;
        int p1 = m - 1;
        int p2 = n - 1;

        while (p1 >= 0 && p2 >= 0) {
            nums1[i--] = nums1[p1] > nums2[p2] ? nums1[p1--] : nums2[p2--];
        }

        while (p2 >= 0) {
            nums1[i--] = nums2[p2--];
        }
    }
}

File: src/main/kotlin/andrew/solution/MergeSortArray.kt
package andrew.solution

/**
 * @author Andrew Tse
 * @date 2019-06-04
 */
fun main(args: Array<String>) {
    val nums1 = intArrayOf(1, 2, 3, 0, 0, 0)
    val nums2 = intArrayOf(2, 5, 6)
    val m = 3
    val n = 3
    merge(nums1, m, nums2, n)
    nums1.forEach { print("$it ") }
}

fun merge(nums1: IntArray, m: Int, nums2: IntArray, n: Int): Unit {
    var i = m + n - 1
    var p1 = m - 1
    var p2 = n - 1
    while (p1 >= 0 && p2 >= 0) {
        nums1[i--] = if (nums1[p1] > nums2[p2]) nums1[p1--] else nums2[p2--]
    }

    while (p2 >= 0) {
        nums1[i--] = nums2[p2--]
    }
}